File: octgnFX/agsxmpp/trunk/samples/VS2005/MiniClient/frmChat.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

using agsXMPP;
using agsXMPP.protocol;
using agsXMPP.protocol.client;
using agsXMPP.Collections;

namespace MiniClient
{
	/// <summary>
	/// 
	/// </summary>
	public class frmChat : System.Windows.Forms.Form
	{
				
		private System.ComponentModel.Container components = null;

		private XmppClientConnection	_connection;
		private Jid						m_Jid;
		private System.Windows.Forms.StatusBar statusBar1;
		private System.Windows.Forms.PictureBox pictureBox1;
		private System.Windows.Forms.Button cmdSend;
		private System.Windows.Forms.RichTextBox rtfSend;
		private System.Windows.Forms.Splitter splitter1;
		private System.Windows.Forms.RichTextBox rtfChat;
		private string					_nickname;

		
		public frmChat(Jid jid, XmppClientConnection con, string nickname)
		{
			m_Jid		= jid;
			_connection = con;
			_nickname	= nickname;
		
			

			InitializeComponent();
			
			this.Text = "Chat with " + nickname;
			
			Util.ChatForms.Add(m_Jid.Bare.ToLower(), this);

			// Setup new Message Callback
            con.MessageGrabber.Add(jid, new BareJidComparer(), new MessageCB(MessageCallback), null);
		}

        public frmChat(Jid jid, XmppClientConnection con, string nickname, bool privateChat)
        {
            m_Jid = jid;
            _connection = con;
            _nickname = nickname;



            InitializeComponent();

            this.Text = "Chat with " + nickname;

            Util.ChatForms.Add(m_Jid.Bare.ToLower(), this);

            // Setup new Message Callback
            if (privateChat)
                con.MessageGrabber.Add(jid, new BareJidComparer(), new MessageCB(MessageCallback), null);
            else
                con.MessageGrabber.Add(jid, new FullJidComparer(), new MessageCB(MessageCallback), null);
        }

		public Jid Jid
		{
			get { return m_Jid; }
			set { m_Jid = value; }
		}
		/// <summary>
		/// 
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if(components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
			
			Util.ChatForms.Remove(m_Jid.Bare.ToLower());
            _connection.MessageGrabber.Remove(m_Jid);
			_connection = null;
		}

		#region Form-Designer Code
		
		private void InitializeComponent()
		{
			this.statusBar1 = new System.Windows.Forms.StatusBar();
			this.pictureBox1 = new System.Windows.Forms.PictureBox();
			this.cmdSend = new System.Windows.Forms.Button();
			this.rtfSend = new System.Windows.Forms.RichTextBox();
			this.splitter1 = new System.Windows.Forms.Splitter();
			this.rtfChat = new System.Windows.Forms.RichTextBox();
			this.SuspendLayout();
			// 
			// statusBar1
			// 
			this.statusBar1.Location = new System.Drawing.Point(0, 214);
			this.statusBar1.Name = "statusBar1";
			this.statusBar1.Size = new System.Drawing.Size(424, 24);
			this.statusBar1.TabIndex = 5;
			// 
			// pictureBox1
			// 
			this.pictureBox1.Dock = System.Windows.Forms.DockStyle.Bottom;
			this.pictureBox1.Location = new System.Drawing.Point(0, 178);
			this.pictureBox1.Name = "pictureBox1";
			this.pictureBox1.Size = new System.Drawing.Size(424, 36);
			this.pictureBox1.TabIndex = 6;
			this.pictureBox1.TabStop = false;
			// 
			// cmdSend
			// 
			this.cmdSend.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Right)));
			this.cmdSend.FlatStyle = System.Windows.Forms.FlatStyle.System;
			this.cmdSend.Location = new System.Drawing.Point(344, 184);
			this.cmdSend.Name = "cmdSend";
			this.cmdSend.Size = new System.Drawing.Size(72, 24);
			this.cmdSend.TabIndex = 7;
			this.cmdSend.Text = "&Send";
			this.cmdSend.Click += new System.EventHandler(this.cmdSend_Click);
			// 
			// rtfSend
			// 
			this.rtfSend.Dock = System.Windows.Forms.DockStyle.Bottom;
			this.rtfSend.Location = new System.Drawing.Point(0, 130);
			this.rtfSend.Name = "rtfSend";
			this.rtfSend.Size = new System.Drawing.Size(424, 48);
			this.rtfSend.TabIndex = 8;
			this.rtfSend.Text = "";
			// 
			// splitter1
			// 
			this.splitter1.Dock = System.Windows.Forms.DockStyle.Bottom;
			this.splitter1.Location = new System.Drawing.Point(0, 122);
			this.splitter1.Name = "splitter1";
			this.splitter1.Size = new System.Drawing.Size(424, 8);
			this.splitter1.TabIndex = 9;
			this.splitter1.TabStop = false;
			// 
			// rtfChat
			// 
			this.rtfChat.Dock = System.Windows.Forms.DockStyle.Fill;
			this.rtfChat.Location = new System.Drawing.Point(0, 0);
			this.rtfChat.Name = "rtfChat";
			this.rtfChat.Size = new System.Drawing.Size(424, 122);
			this.rtfChat.TabIndex = 10;
			this.rtfChat.Text = "";
			// 
			// frmChat
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(424, 238);
			this.Controls.Add(this.rtfChat);
			this.Controls.Add(this.splitter1);
			this.Controls.Add(this.rtfSend);
			this.Controls.Add(this.cmdSend);
			this.Controls.Add(this.pictureBox1);
			this.Controls.Add(this.statusBar1);
			this.Name = "frmChat";
			this.Text = "frmChat";
			this.ResumeLayout(false);

		}
		#endregion

		private void OutgoingMessage(agsXMPP.protocol.client.Message msg)
		{
			rtfChat.SelectionColor = Color.Blue;
			rtfChat.AppendText("Me said: ");
			rtfChat.SelectionColor = Color.Black;
			rtfChat.AppendText(msg.Body);
			rtfChat.AppendText("\r\n");
		}

		public void IncomingMessage(agsXMPP.protocol.client.Message msg)
		{
			rtfChat.SelectionColor = Color.Red;
			rtfChat.AppendText(_nickname + " said: ");
			rtfChat.SelectionColor = Color.Black;
			rtfChat.AppendText(msg.Body);
			rtfChat.AppendText("\r\n");
		}

		private void cmdSend_Click(object sender, System.EventArgs e)
		{
			agsXMPP.protocol.client.Message msg = new agsXMPP.protocol.client.Message();

			msg.Type	= MessageType.chat;
			msg.To		= m_Jid;
			msg.Body	= rtfSend.Text;
			
			_connection.Send(msg);
			OutgoingMessage(msg);
			rtfSend.Text = "";
		}

		private void MessageCallback(object sender, agsXMPP.protocol.client.Message msg, object data)
		{
            if (InvokeRequired)
            {
                // Windows Forms are not Thread Safe, we need to invoke this :(
                // We're not in the UI thread, so we need to call BeginInvoke				
                BeginInvoke(new MessageCB(MessageCallback), new object[] { sender, msg, data });
                return;
            }
            
            if (msg.Body != null)
			    IncomingMessage(msg);
		}

		
	}
}

File: octgnFX/agsxmpp/trunk/samples/VS2003/MiniClientVB/frmChat.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports agsXMPP
Imports agsXMPP.protocol
Imports agsXMPP.protocol.client
Imports agsXMPP.Collections
Namespace MiniClient

	Public Class frmChat
	Inherits System.Windows.Forms.Form
		Private components As System.ComponentModel.Container = Nothing
		Private _connection As XmppClientConnection
		Private m_Jid As Jid
		Private _nickname As String
		Private rtfChat As System.Windows.Forms.RichTextBox
		Private splitter1 As System.Windows.Forms.Splitter
		Private cmdSend As System.Windows.Forms.Button
		Private statusBar1 As System.Windows.Forms.StatusBar
		Private rtfSend As System.Windows.Forms.RichTextBox
		Private pictureBox1 As System.Windows.Forms.PictureBox
		
		Public Sub New(ByVal jid As Jid, ByVal con As XmppClientConnection, ByVal nickname As String)
			m_Jid = jid
			_connection = con
			_nickname = nickname
			InitializeComponent
			Me.Text = "Chat with " + nickname
			Util.Forms.Add(m_Jid.Bare.ToLower, Me)
            con.MessageGrabber.Add(jid, New BareJidComparer, AddressOf MessageCallback, Nothing)
		End Sub

		Public Property Jid() As Jid
			Get
				Return m_Jid
			End Get
			Set
				m_Jid = value
			End Set
		End Property

		Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If Not (components Is Nothing) Then
					components.Dispose
				End If
			End If
			MyBase.Dispose(disposing)
			Util.Forms.Remove(m_Jid.Bare.ToLower)
            _connection.MessageGrabber.Remove(m_Jid)
			_connection = Nothing
		End Sub

		Private Sub InitializeComponent()
			Me.pictureBox1 = New System.Windows.Forms.PictureBox
			Me.rtfSend = New System.Windows.Forms.RichTextBox
			Me.statusBar1 = New System.Windows.Forms.StatusBar
			Me.cmdSend = New System.Windows.Forms.Button
			Me.splitter1 = New System.Windows.Forms.Splitter
			Me.rtfChat = New System.Windows.Forms.RichTextBox
			Me.SuspendLayout
			'
			'pictureBox1
			'
			Me.pictureBox1.Dock = System.Windows.Forms.DockStyle.Bottom
			Me.pictureBox1.Location = New System.Drawing.Point(0, 178)
			Me.pictureBox1.Name = "pictureBox1"
			Me.pictureBox1.Size = New System.Drawing.Size(424, 36)
			Me.pictureBox1.TabIndex = 6
			Me.pictureBox1.TabStop = false
			'
			'rtfSend
			'
			Me.rtfSend.Dock = System.Windows.Forms.DockStyle.Bottom
			Me.rtfSend.Location = New System.Drawing.Point(0, 130)
			Me.rtfSend.Name = "rtfSend"
			Me.rtfSend.Size = New System.Drawing.Size(424, 48)
			Me.rtfSend.TabIndex = 8
			Me.rtfSend.Text = ""
			'
			'statusBar1
			'
			Me.statusBar1.Location = New System.Drawing.Point(0, 214)
			Me.statusBar1.Name = "statusBar1"
			Me.statusBar1.Size = New System.Drawing.Size(424, 24)
			Me.statusBar1.TabIndex = 5
			'
			'cmdSend
			'
			Me.cmdSend.Anchor = CType((System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Right),System.Windows.Forms.AnchorStyles)
			Me.cmdSend.FlatStyle = System.Windows.Forms.FlatStyle.System
			Me.cmdSend.Location = New System.Drawing.Point(344, 184)
			Me.cmdSend.Name = "cmdSend"
			Me.cmdSend.Size = New System.Drawing.Size(72, 24)
			Me.cmdSend.TabIndex = 7
			Me.cmdSend.Text = "&Send"
			AddHandler Me.cmdSend.Click, AddressOf Me.cmdSend_Click
			'
			'splitter1
			'
			Me.splitter1.Dock = System.Windows.Forms.DockStyle.Bottom
			Me.splitter1.Location = New System.Drawing.Point(0, 122)
			Me.splitter1.Name = "splitter1"
			Me.splitter1.Size = New System.Drawing.Size(424, 8)
			Me.splitter1.TabIndex = 9
			Me.splitter1.TabStop = false
			'
			'rtfChat
			'
			Me.rtfChat.Dock = System.Windows.Forms.DockStyle.Fill
			Me.rtfChat.Location = New System.Drawing.Point(0, 0)
			Me.rtfChat.Name = "rtfChat"
			Me.rtfChat.Size = New System.Drawing.Size(424, 122)
			Me.rtfChat.TabIndex = 10
			Me.rtfChat.Text = ""
			'
			'frmChat
			'
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 14)
			Me.ClientSize = New System.Drawing.Size(424, 238)
			Me.Controls.Add(Me.rtfChat)
			Me.Controls.Add(Me.splitter1)
			Me.Controls.Add(Me.rtfSend)
			Me.Controls.Add(Me.cmdSend)
			Me.Controls.Add(Me.pictureBox1)
			Me.Controls.Add(Me.statusBar1)
			Me.Name = "frmChat"
			Me.Text = "frmChat"
			Me.ResumeLayout(false)
		End Sub

		Private Sub OutgoingMessage(ByVal msg As agsXMPP.protocol.client.Message)
			rtfChat.SelectionColor = Color.Blue
			rtfChat.AppendText("Me said: ")
			rtfChat.SelectionColor = Color.Black
			rtfChat.AppendText(msg.Body)
			rtfChat.AppendText("" & Microsoft.VisualBasic.Chr(13) & "" & Microsoft.VisualBasic.Chr(10) & "")
		End Sub

		Public Sub IncomingMessage(ByVal msg As agsXMPP.protocol.client.Message)
			rtfChat.SelectionColor = Color.Red
			rtfChat.AppendText(_nickname + " said: ")
			rtfChat.SelectionColor = Color.Black
			rtfChat.AppendText(msg.Body)
			rtfChat.AppendText("" & Microsoft.VisualBasic.Chr(13) & "" & Microsoft.VisualBasic.Chr(10) & "")
		End Sub

		Private Sub cmdSend_Click(ByVal sender As Object, ByVal e As System.EventArgs)
			Dim msg As agsXMPP.protocol.client.Message = New agsXMPP.protocol.client.Message
			msg.Type = MessageType.chat
			msg.To = m_Jid
			msg.Body = rtfSend.Text
			_connection.Send(msg)
			OutgoingMessage(msg)
			rtfSend.Text = ""
		End Sub

		Private Sub MessageCallback(ByVal sender As Object, ByVal msg As agsXMPP.protocol.client.Message, ByVal data As Object)
			IncomingMessage(msg)
		End Sub
	End Class 
End Namespace
